core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -116,15 +121,26 @@ public void execute(SearchContext searchContext) throws QueryPhaseExecutionExcep
         // here to make sure it happens during the QUERY phase
         aggregationPhase.preProcess(searchContext);
 
-        searchContext.queryResult().searchTimedOut(false);
+        boolean rescore = execute(searchContext, searchContext.searcher());
+
+        if (rescore) { // only if we do a regular search
+            rescorePhase.execute(searchContext);
+        }
+        suggestPhase.execute(searchContext);
+        aggregationPhase.execute(searchContext);
+    }
+
+    // extracted to a pkg-private static method so that this can be tested
+    static boolean execute(SearchContext searchContext, final IndexSearcher searcher) throws QueryPhaseExecutionException {
```

Comment:
Maybe switch to Javadoc? Its easier to read and it'd prompt you to write some docs on what that return boolean means.